src/sumlog.jl

```diff
@@ -0,0 +1,36 @@
+using IrrationalConstants: logtwo
+
+"""
+$(SIGNATURES)
+
+Compute `sum(log.(X))`.
+
+`sum(log.(X))` can be evaluated much more quickly as `sum(log, X)`. However,
+this still requires computing `log` for each element of `X`. 
+
+`sumlog(X)` can be faster still, especially an the length of `X` increases.
+
+This works by representing the `j`th element of `X` as `xⱼ = aⱼ * 2 ^ bⱼ`,
+allowing us to write
+
+    ∑ⱼ log(xⱼ) = log(∏ⱼ aⱼ) + log(2) * ∑ⱼ bⱼ
+
+Since `log(2)` is constant, `sumlog` only requires a single `log` evaluation.
+"""
+function sumlog(x::AbstractArray{T}) where {T} 
+    sig = one(T) 
+    ex = zero(exponent(one(T)))
+    bound = floatmax(T) / 2 
+    for xj in x 
+        sig *= significand(xj)
+        ex += exponent(xj) 
```

Comment:
Maybe one could use
```suggestion
function sumlog(x::AbstractArray{<:Real})
    T = float(eltype(x))
    sig = one(T) 
    ex = zero(exponent(sig))
    bound = floatmax(T) / 2 
    for xj in x
        float_xj = float(xj)
        sig *= significand(float_xj)
        ex += exponent(float_xj) 
```

Reply:
Maybe even `<:AbstractFloat`? Then we could have a fall-back `sumlog(x) = sum(log, x)` so all cases should be covered

Reply:
I thought about this but the main disadvantage would be that `Int`s, `Rational`s etc all would not benefit from the speedup even though their result is a floating point number.

Reply:
Right, so having `sumlog(x) = sum(log, x)` would make them work, allowing the same code to work better across types. Maybe I'm misinderstanding you?

Reply:
I mean it would work but it would not be optimized. Even though I think there's no reason why it shouldn't, and rather it would be surprising that `sumlog(ones(10_000))` and `sumlog(ones(Int, 10_000))` would be have differently. With the suggestion above both would only use one log evaluation.

Reply:
I see. It looks like this does work better than the fallback `log(sum, x)`.

One concern I have with this is that `Real` is for symbolic types. I think I see a way to test for that without depending on them. I'll resolve this then try that so we can discuss separately